### PR TITLE
add tag_depth=1 support to add_samples_from_path

### PR DESCRIPTION
**Pull request by masakljun**
## What has changed and why?

Added support for automatically creating sample tags when adding data from a folder structure.
When calling add_samples_from_path with tag_depth=1, the function now creates a tag for each parent directory and assigns new samples to those tags.
This enables users to organize data imported from directory-based datasets.
Related to https://github.com/lightly-ai/lightly-studio/issues/7#issuecomment-3465939257
## How has it been tested?
Added 3 tests to test_dataset.py

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [ ] Not needed (internal change)

### CODE REVIEW

File: lightly_studio/src/lightly_studio/core/dataset.py [hunk 1]
```diff
@@ -288,6 +303,36 @@ def add_samples_from_path(
...
+
+            # Fetch newly created samples to get their file paths
+            newly_created_samples = (
+                self.query()
+                .match(SampleField.id.is_in(created_sample_ids))
+                .to_list()
+            )
+
+            for sample in newly_created_samples:
+                parent_dir = Path(sample.file_path_abs).parent.name.strip()
```
Comment (MalteEbner) [inline lightly_studio/src/lightly_studio/core/dataset.py]: This will not get the depth=1, but the depth=last directory.

E.g. if `path = "/my/path"` and `sample.file_path_abs="/my/path/dir_depth_1/dir_depth_2/img.jpg`, then `parent_dir="dir_depth_2"`.
The solution is to rather use a logic similar to `Path(sample.file_path_abs).relative_to(path).split("/")[0]`.

You also need to handle the case of an image being in the path directly, e.g. `path = "/my/path"` and `sample.file_path_abs="/my/path/img.jpg` should not lead to an error.

File: lightly_studio/src/lightly_studio/core/dataset.py [hunk 2]
```diff
@@ -288,6 +303,36 @@ def add_samples_from_path(
...
+                self.query()
+                .match(SampleField.id.is_in(created_sample_ids))
+                .to_list()
+            )
+
+            for sample in newly_created_samples:
+                parent_dir = Path(sample.file_path_abs).parent.name.strip() 
+                if not parent_dir:
+                    parent_dir = "root"  
+                parent_dir_to_sample_ids[parent_dir].append(sample.sample_id)
```
Comment (MalteEbner) [inline lightly_studio/src/lightly_studio/core/dataset.py]: The current implementation is contrary to the GitHub issue https://github.com/lightly-ai/lightly-studio/issues/7#issuecomment-3465939257, which tells to not set a tag when an image is in the root directory.

```suggestion
                if parent_dir:
                    parent_dir_to_sample_ids[parent_dir].append(sample.sample_id)
```

File: lightly_studio/src/lightly_studio/core/dataset.py [hunk 3]
```diff
@@ -288,6 +303,36 @@ def add_samples_from_path(
             image_paths=image_paths,
         )
 
+        if tag_depth == 1 and created_sample_ids:
+            print(f"Adding directory tags to {len(created_sample_ids)} new samples.")
```
Comment (MalteEbner) [inline lightly_studio/src/lightly_studio/core/dataset.py]: Can you move this whole logic into a separate function? E.g.

`def tag_images_based_on_directories(image_root: PathLike, image_ids: Sequence[UUID])`

File: lightly_studio/src/lightly_studio/core/dataset.py [hunk 4]
```diff
@@ -288,11 +304,68 @@ def add_samples_from_path(
...
+
+        try:
+            input_path_abs = Path(input_path).absolute()
+        except Exception as e:
+            print(f"Warning: Could not resolve absolute path for {input_path}. Skipping tagging. Error: {e}")
+            return
+
+        newly_created_samples = (
+            self.query()
+            .match(SampleField.id.is_in(sample_ids))
```
Comment (masakljun) [inline lightly_studio/src/lightly_studio/core/dataset.py]: Hey @MalteEbner , a quick question/thought:

In my function I need to get the ids of the new samples, so I can then get the correct file paths for the creation of the new tags.

Initially I thought I could just get the ids from the file lister at the start of `add_samples_from_path()`. However, the `image_paths` list and the `created_sample_ids` are not necessarily a 1 to 1 match (if some images don't successfully get turned into samples).

Therefore, I think one option is to re-query to get the correct `file_path_abs` for each sample id that was actually created. 

While the other option is to add `sample_id` to the `SampleField` class. This seems like the cleaner option, as it makes SampleField more complete. Then I could do  .match(SampleField.sample_id.is_in(sample_ids)).

What do you think?

File: lightly_studio/src/lightly_studio/core/dataset.py [hunk 5]
```diff
@@ -288,11 +305,68 @@ def add_samples_from_path(
...
 
+    def _tag_samples_by_directory(
+        self,
+        input_path: PathLike,
+        sample_ids: list[UUID],
+    ) -> None:
+        """Tags samples based on their first-level subdirectory relative to input_path."""
+
+        print(f"Adding directory tags to {len(sample_ids)} new samples.")
+        parent_dir_to_sample_ids = defaultdict(list)
```
Comment (MalteEbner) [inline lightly_studio/src/lightly_studio/core/dataset.py]: Minor: Can you move the `parent_dir_to_sample_ids` var closer to where it is used? E.g. directly before the for-loop?

File: lightly_studio/src/lightly_studio/core/dataset.py [hunk 6]
```diff
@@ -288,11 +305,68 @@ def add_samples_from_path(
...
+
+        try:
+            input_path_abs = Path(input_path).absolute()
+        except Exception as e:
```
Comment (MalteEbner) [inline lightly_studio/src/lightly_studio/core/dataset.py]: In our [python code style notion docs](https://www.notion.so/getlightly/Python-code-style-290e575bc2134c3eb4c67d28e350d954), it links to the google style guide, which covers exceptions. 

Some relevant part of it:

> Never use catch-all except: statements, or catch Exception or StandardError

In this case: I would remove the try-catch completely and just let users run into the error.

File: lightly_studio/src/lightly_studio/core/dataset.py [hunk 7]
```diff
@@ -288,11 +305,68 @@ def add_samples_from_path(
...
+
+                if len(relative_path.parts) > 1:
+                    tag_name = relative_path.parts[0].strip()
+                    if tag_name:
+                        parent_dir_to_sample_ids[tag_name].append(sample.sample_id)
+
+            except ValueError:
+                print(f"Warning: Sample {sample.file_name} is not in the input path. Skipping tagging.")
+            except Exception as e:
+                print(f"Warning: Error processing tag for {sample.file_name}: {e}. Skipping.")
```
Comment (MalteEbner) [inline lightly_studio/src/lightly_studio/core/dataset.py]: Similarly to above, I would remove this exception catching completely.

File: lightly_studio/src/lightly_studio/core/dataset.py [hunk 8]
```diff
@@ -288,11 +305,68 @@ def add_samples_from_path(
...
+            print(f"Warning: Could not resolve absolute path for {input_path}. Skipping tagging. Error: {e}")
+            return
+
+        # todo: this doesnt work ATM since sample_id is not attribute of SampleField
+        newly_created_samples = (
+            self.query()
+            .match(SampleField.sample_id.is_in(sample_ids)) 
+            .to_list()
+        )
+        for sample in newly_created_samples:
```
Comment (MalteEbner) [inline lightly_studio/src/lightly_studio/core/dataset.py]: The self.query is the public interface. Internally, we can use the resolvers directly. E.g. we can use
[sample_resolver.get_many_by_id](https://github.com/lightly-ai/lightly-studio/blob/e534887e58781421641b4d713a73b180317a0292/lightly_studio/src/lightly_studio/resolvers/sample_resolver.py#L40)

File: lightly_studio/tests/core/test_dataset.py [hunk 9]
```diff
@@ -356,6 +358,78 @@ def test_compute_typicality_metadata(
...
+        img.save(tmp_path / "site_1" / "deep_dir" / "img2.png")
+        (tmp_path / " site_2 ").mkdir()
+        img.save(tmp_path / " site_2 " / "img3.png")
+
+        # Run with tag_depth=1
+        dataset = Dataset.create()
+        dataset.add_samples_from_path(tmp_path, tag_depth=1, embed=False)
+
+        # Order by filename to have a consistent order for checking
+        samples = dataset.query().order_by(SampleField.file_name).to_list()
```
Comment (MalteEbner) [inline lightly_studio/tests/core/test_dataset.py]: We don't need a consistent filename order, as the samples are only used for creating a dictionary.
```suggestion
        # Order by filename to have a consistent order for checking
        samples = dataset.query().to_list()
```

File: lightly_studio/tests/core/test_dataset.py [hunk 10]
```diff
@@ -356,6 +358,78 @@ def test_compute_typicality_metadata(
...
+        assert len(samples) == 4
+
+        # Create a map of filename -> {set of tag names} for easy checking
+        tag_map = {s.file_name: {tag.name for tag in s.tags} for s in samples}
```
Comment (MalteEbner) [inline lightly_studio/tests/core/test_dataset.py]: We usually name dictionaries in the format of `f"{key_name}_to_{value_name}"`.

```suggestion
        # Create a map of filename -> {set of tag names} for easy checking
        sample_filename_to_tags = {s.file_name: {tag.name for tag in s.tags} for s in samples}
```

File: lightly_studio/tests/core/test_dataset.py [hunk 11]
```diff
@@ -356,6 +358,78 @@ def test_compute_typicality_metadata(
...
         assert samples[2].metadata["typicality"] == pytest.approx(0.3853082)
 
+    def test_add_samples_from_path_tag_depth_invalid(
+        self,
+        patch_dataset: None,  # noqa: ARG002
+        tmp_path: Path,
+    ) -> None:
+        """Tests that tag_depth > 1 raises an error."""
+        dataset = Dataset.create(name="test_dataset")
+        with pytest.raises(NotImplementedError):
```
Comment (MalteEbner) [inline lightly_studio/tests/core/test_dataset.py]: When asserting errors, we usually assert parts of the error message using the `match` argument.

```suggestion
        with pytest.raises(NotImplementedError, match=r"tag_depth > 1 is not yet implemented for add_samples_from_path"):
```

Not sure if this works, you might need to escape some special chars of the match argument. Best try it out.

File: lightly_studio/tests/core/test_dataset.py [hunk 12]
```diff
@@ -356,6 +358,78 @@ def test_compute_typicality_metadata(
...
+    ) -> None:
+        """Tests the default behavior (tag_depth=0) adds samples but no tags."""
+                
+        # Create a valid image
+        img = Image.new("RGB", (10, 10), color="red")
+        
+        img.save(tmp_path / "root_img.png")
+        
+        (tmp_path / "site_1").mkdir()
+        img.save(tmp_path / "site_1" / "img1.png")
```
Comment (MalteEbner) [inline lightly_studio/tests/core/test_dataset.py]: Please use the test helper function [create_images](https://github.com/lightly-ai/lightly-studio/blob/e534887e58781421641b4d713a73b180317a0292/lightly_studio/tests/helpers_resolvers.py#L127) instead.

File: lightly_studio/tests/core/test_dataset.py [hunk 13]
```diff
@@ -356,6 +358,78 @@ def test_compute_typicality_metadata(
...
+        """Tests that tag_depth=1 correctly tags samples based on directory structure."""
+        # Setup file structure with all edge cases
+        img = Image.new("RGB", (10, 10), color="blue")
+        img.save(tmp_path / "root_img.png")
+        (tmp_path / "site_1").mkdir()
+        img.save(tmp_path / "site_1" / "img1.png")
+        (tmp_path / "site_1" / "deep_dir").mkdir(parents=True)
+        img.save(tmp_path / "site_1" / "deep_dir" / "img2.png")
+        (tmp_path / " site_2 ").mkdir()
+        img.save(tmp_path / " site_2 " / "img3.png")
```
Comment (MalteEbner) [inline lightly_studio/tests/core/test_dataset.py]: Please use the test helper function [create_images](https://github.com/lightly-ai/lightly-studio/blob/e534887e58781421641b4d713a73b180317a0292/lightly_studio/tests/helpers_resolvers.py#L127) instead.

File: lightly_studio/src/lightly_studio/core/dataset.py (hunk 4 above)
Comment (MalteEbner) [reply lightly_studio/src/lightly_studio/core/dataset.py]: > In my function I need to get the ids of the new samples, so I can then get the correct file paths for the creation of the new tags. However, the image_paths list and the created_sample_ids are not necessarily a 1 to 1 match (if some images don't successfully get turned into samples).

Yes, you got that correctly. The goal is to add the tag to all samples that were actually created. Thus you can just use the `created_sample_ids` as you already do in your PR.

Comment (MalteEbner) [reply lightly_studio/src/lightly_studio/core/dataset.py]: > While the other option is to add sample_id to the SampleField class. This seems like the cleaner option, as it makes SampleField more complete. Then I could do .match(SampleField.sample_id.is_in(sample_ids)).

See https://github.com/lightly-ai/lightly-studio/pull/84#discussion_r2494580096

File: lightly_studio/src/lightly_studio/core/dataset.py [hunk 14]
```diff
@@ -282,20 +291,69 @@ def add_samples_from_path(
                 path=str(path), allowed_extensions=allowed_extensions_set
             )
         )
+
+        if tag_depth > 1:
+            raise NotImplementedError(
+                "tag_depth > 1 is not yet implemented for add_samples_from_path."
+            )
```
Comment (MalteEbner) [inline lightly_studio/src/lightly_studio/core/dataset.py]: I thought about the location of this. I would move this into `_tag_samples_by_directory`.
Then we can call `_tag_samples_by_directory` from `add_samples_from_coco` and `add_samples_from_yolo` in the future. 

See also the comment one below.

File: lightly_studio/src/lightly_studio/core/dataset.py [hunk 15]
```diff
@@ -282,20 +291,69 @@ def add_samples_from_path(
...
 
-        # Process images.
+        # Process images
         created_sample_ids = add_samples.load_into_dataset_from_paths(
             session=self.session,
             dataset_id=self.dataset_id,
             image_paths=image_paths,
         )
 
+        if tag_depth == 1 and created_sample_ids:
```
Comment (MalteEbner) [inline lightly_studio/src/lightly_studio/core/dataset.py]: Let's move the validation of `tag_depth` into `_tag_samples_by_directory`.
If tag_depth=0: early return
if tag_depth>1: raise NotImplementedError
else: normal path

```suggestion
        if created_sample_ids:
```

See also the comment one above.

File: lightly_studio/src/lightly_studio/core/dataset.py [hunk 16]
```diff
@@ -282,20 +291,69 @@ def add_samples_from_path(
...
+        newly_created_samples = [Sample(inner=image) for image in newly_created_images]
+
+        print(f"Adding directory tags to {len(sample_ids)} new samples.")
+        parent_dir_to_sample_ids: defaultdict[str, list[UUID]] = defaultdict(list)
+        for sample in newly_created_samples:
+            sample_path_abs = Path(sample.file_path_abs)
+            relative_path = sample_path_abs.relative_to(input_path_abs)
+
+            if len(relative_path.parts) > 1:
+                tag_name = relative_path.parts[0].strip()
```
Comment (MalteEbner) [inline lightly_studio/src/lightly_studio/core/dataset.py]: I  would leave the stripping away. If users actually have a space in the path, e.g. `/my/ dir/` and `/my/dir /` and `/my/dir/`, they probably care about preserving that space.

This makes it also possible for users to get the directory name from the tag name, i.e. users may want to  do something like the following. When stripping spaces, it will fail instead.

```
for tag = dataset.get_tags() # get_tags() is a placeholder, it does not exist
     directory = Path(input_path) / tag.name
     os.copy(directory, new_path / tag.name)
```


```suggestion
                tag_name = relative_path.parts[0]
```

File: lightly_studio/src/lightly_studio/core/dataset.py [hunk 17]
```diff
@@ -282,20 +291,69 @@ def add_samples_from_path(
...
         if embed:
             _generate_embeddings(
                 session=self.session, dataset_id=self.dataset_id, sample_ids=created_sample_ids
             )
 
+    def _tag_samples_by_directory(
+        self,
+        input_path: PathLike,
+        sample_ids: list[UUID],
+    ) -> None:
```
Comment (MalteEbner) [inline lightly_studio/src/lightly_studio/core/dataset.py]: Optional, idea:
Sorry for only thinking about this now, I just had a closer look at the overall directory structure.

What do you think about moving this function to [add_samples.py](https://github.com/lightly-ai/lightly-studio/blob/main/lightly_studio/src/lightly_studio/core/add_samples.py) instead?

Feel free to challenge that idea and give me reasons why not to do it ;)

File: lightly_studio/tests/core/test_dataset.py [hunk 18]
```diff
@@ -356,6 +356,108 @@ def test_compute_typicality_metadata(
...
+        self,
+        patch_dataset: None,  # noqa: ARG002
+    ) -> None:
+        """Tests that tag_depth > 1 raises an error."""
+        dataset = Dataset.create(name="test_dataset")
+        with pytest.raises(
+            NotImplementedError,
+            match="tag_depth > 1 is not yet implemented for add_samples_from_path",
+        ):
+            dataset.add_samples_from_path(".", tag_depth=2)
```
Comment (MalteEbner) [inline lightly_studio/tests/core/test_dataset.py]: All 3 tests currently call `dataset.add_samples_from_path` and not `_tag_samples_by_directory`. Thus they are more integration tests than unittests.

I would change them to call `_tag_samples_by_directory` instead. 
From a testing point of view, this makes sure that only and exactly the relevant part is tested.
From a test implementation point of view, it allows to leave away all the the mocking of `lightly_studio.dataset.fsspec_lister.iter_files_from_path`, the files do not need to be on disk anymore. It is sufficient if the samples are in the DB, which happens with `create_images(`.

**Review by MalteEbner — COMMENTED** [review]
The logic of `_tag_samples_by_directory` is now almost correctly, only some minor improvement ideas.

Regarding the tests: After moving the `tag_depth` logic into `_tag_samples_by_directory`, the tests should only test `_tag_samples_by_directory` and not the whole `add_samples_from_path` anymore. See https://github.com/lightly-ai/lightly-studio/pull/84/files#r2502176747

File: lightly_studio/src/lightly_studio/core/dataset.py (hunk 17 above)
Comment (masakljun) [reply lightly_studio/src/lightly_studio/core/dataset.py]: Thanks for the suggestion. I'm not familiar with the codebase yet, but I had a look through `add_samples.py` and `dataset.py`, and what you said makes sense to me.

It feels cleaner to keep the Dataset class as the main interface for users and move the actual tagging logic next to the other functions that handle adding data to the database.

So yeah I think it makes sense to move as it will be more organized that way

File: lightly_studio/tests/core/test_dataset.py (hunk 18 above)
Comment (masakljun) [reply lightly_studio/tests/core/test_dataset.py]: Good idea, thanks!

**Comment by masakljun** [comment]
> The logic of `_tag_samples_by_directory` is now almost correctly, only some minor improvement ideas.
> 
> Regarding the tests: After moving the `tag_depth` logic into `_tag_samples_by_directory`, the tests should only test `_tag_samples_by_directory` and not the whole `add_samples_from_path` anymore. See https://github.com/lightly-ai/lightly-studio/pull/84/files#r2502176747

Thank you again for your review and great feedback. I now changed the logic but have some issues still with tests not going through, will try to resolve and then push the changes.

File: lightly_studio/tests/core/test_add_samples.py [hunk 19]
```diff
@@ -228,6 +235,112 @@ def test_create_label_map(db_session: Session) -> None:
...
+
+    images_to_create = [
+        ImageStub(path=mock_image_paths[0]),
+        ImageStub(path=mock_image_paths[1]),
+    ]
+    created_images = create_images(
+        db_session=db_session,
+        dataset_id=dataset_table.dataset_id,
+        images=images_to_create,
+    )
```
Comment (MalteEbner) [inline lightly_studio/tests/core/test_add_samples.py:268]: This can be shortened a lot, no need for all the temporary variables.

```suggestion
    dataset_table = create_dataset(db_session, "test_dataset")
    mock_root_path = "/mock/path"
    created_images = create_images(
        db_session=db_session,
        dataset_id=dataset_table.dataset_id,
        images=[
            ImageStub(path=f"{mock_root_path}/root_img.png"),
            ImageStub(path=f"{mock_root_path}/site_1/img1.png")
        ],
    )
```

File: lightly_studio/tests/core/test_add_samples.py [hunk 20]
```diff
@@ -228,6 +235,112 @@ def test_create_label_map(db_session: Session) -> None:
...
+        ImageStub(path=mock_image_paths[0]),
+        ImageStub(path=mock_image_paths[1]),
+        ImageStub(path=mock_image_paths[2]),
+        ImageStub(path=mock_image_paths[3]),
+    ]
+    created_images = create_images(
+        db_session=db_session,
+        dataset_id=dataset_table.dataset_id,
+        images=images_to_create,
+    )
```
Comment (MalteEbner) [inline lightly_studio/tests/core/test_add_samples.py:304]: This can be shortened a lot, see the comment above.

File: lightly_studio/tests/core/test_add_samples.py [hunk 21]
```diff
@@ -14,11 +16,16 @@
...
 from lightly_studio.core import add_samples
 from lightly_studio.models.image import ImageCreate
 from lightly_studio.models.sample import SampleTable
 from lightly_studio.resolvers import caption_resolver, image_resolver
-from tests.helpers_resolvers import create_dataset
+from tests.helpers_resolvers import (
+    ImageStub,
+    create_dataset,
+    create_images,
+)
```
Comment (MalteEbner) [inline lightly_studio/tests/core/test_add_samples.py:27]: Code style: We don't import functions directly, but rather their file. Then call `helpers_resolver.create_dataset` and `helpers_resolvers.create_image`. 

```suggestion
from tests.helpers_resolvers import (
    ImageStub,
)
from tests import helpers_resolvers
```

File: lightly_studio/tests/core/test_add_samples.py [hunk 22]
```diff
@@ -341,6 +325,35 @@ def test_tag_samples_by_directory_tag_depth_1(
...
+    mocker.patch("lightly_studio.core.dataset._generate_embeddings")
+    spy_tagger = mocker.spy(add_samples, "tag_samples_by_directory")
+
+    dataset_table = helpers_resolvers.create_dataset(db_session, "test_dataset")
+    dataset = Dataset(dataset=dataset_table)
+    dataset.session = db_session
+
+    _create_sample_images([tmp_path / "image1.jpg"])
+
+    dataset.add_samples_from_path(path=str(tmp_path), tag_depth=0)
```
Comment (MalteEbner) [inline lightly_studio/tests/core/test_add_samples.py]: I did not test it, but is setting `embed=False` sufficient, so that `mocker.patch("lightly_studio.core.dataset._generate_embeddings")` is not needed?

```suggestion
    spy_tagger = mocker.spy(add_samples, "tag_samples_by_directory")

    dataset_table = helpers_resolvers.create_dataset(db_session, "test_dataset")
    dataset = Dataset(dataset=dataset_table)
    dataset.session = db_session

    _create_sample_images([tmp_path / "image1.jpg"])

    dataset.add_samples_from_path(path=str(tmp_path), tag_depth=0, embed=False)
```

Feel free to resolve this comment yourself if it does not work.

File: lightly_studio/tests/core/test_add_samples.py [hunk 23]
```diff
@@ -341,6 +325,35 @@ def test_tag_samples_by_directory_tag_depth_1(
     assert sample_filename_to_tags["root_img.png"] == set()
 
 
+def test_add_samples_from_path_calls_tag_samples_by_directory(
```
Comment (MalteEbner) [inline lightly_studio/tests/core/test_add_samples.py]: Major: This belongs into `lightly_studio/tests/core/test_dataset__path.py`, as it is testing a function in `dataset.py`, not in `add_samples.py`.
As `test_dataset.py` would become to large if we put all tests into it, we have split it into different test files, including `test_dataset__path.py` for testing `Dataset.add_samples_from_path`.

You can move it easily in VSCode via right-clicking on the function name -> Refactor -> Move symbol to. This will also move all imports.

File: lightly_studio/tests/core/test_add_samples.py [hunk 24]
```diff
@@ -341,6 +325,35 @@ def test_tag_samples_by_directory_tag_depth_1(
...
+    The add_samples.tag_samples_by_directory helper.
+    """
+    mocker.patch("lightly_studio.core.dataset._generate_embeddings")
+    spy_tagger = mocker.spy(add_samples, "tag_samples_by_directory")
+
+    dataset_table = helpers_resolvers.create_dataset(db_session, "test_dataset")
+    dataset = Dataset(dataset=dataset_table)
+    dataset.session = db_session
+
+    _create_sample_images([tmp_path / "image1.jpg"])
```
Comment (MalteEbner) [inline lightly_studio/tests/core/test_add_samples.py]: Minor: This can be done more easily, see https://github.com/lightly-ai/lightly-studio/blob/e716cbaa06b9276389695cc8a6138597aacc059a/lightly_studio/tests/core/test_dataset__path.py#L177-L179

Comment (masakljun) [reply lightly_studio/tests/core/test_add_samples.py]: Thanks for the suggestion!

However, I tried using `Dataset.create()`, but it looks like it causes an `IntegrityError` in this specific test. It seems `Dataset.create()` uses the app's persistent session, but this test requires everything to use the isolated `db_session` fixture.

So in the end I now changed back to my original, as this then causes the test to pass.

File: lightly_studio/tests/core/test_add_samples.py (hunk 22 above)
Comment (masakljun) [reply lightly_studio/tests/core/test_add_samples.py]: Ah, yes, you're right, using `embed=False` is much cleaner. Thanks!

File: lightly_studio/tests/core/test_dataset__path.py [hunk 25]
```diff
@@ -189,3 +193,30 @@ def _create_sample_images(image_paths: list[Path]) -> None:
...
+
+    dataset.add_samples_from_path(path=str(tmp_path), tag_depth=0, embed=False)
+
+    spy_tagger.assert_called_once_with(
+        session=db_session,
+        dataset_id=dataset.dataset_id,
+        input_path=str(tmp_path),
+        sample_ids=mocker.ANY,
+        tag_depth=0,
+    )
```
Comment (MalteEbner) [inline lightly_studio/tests/core/test_dataset__path.py]: As this is a test for a method and not a class-less function, it should be a method of the `TestDataset` class earlier in this file. Thus just move it there. See `test_dataset_add_samples_from_path__dont_embed` in the same file a few lines earlier, it can go directly below it, while still being a method. Note that this means that the `self` argument needs to be added to `test_add_samples_from_path_calls_tag_samples_by_directory`

Comment (MalteEbner) [reply lightly_studio/tests/core/test_dataset__path.py]: Feel free to implement this change and then resolve the comment yourself.
